cpp/src/arrow/compute/util.cc

```diff
@@ -58,6 +58,13 @@ void TempVectorStack::release(int id, uint32_t num_bytes) {
   --num_vectors_;
 }
 
+void TempVectorStack::CheckAllocSizeValid(int64_t estimate_alloc_size) {
+  ARROW_DCHECK_LE(estimate_alloc_size, buffer_size_)
```

Comment:
I think that we should receive additional allocation size instead of total new allocation size here: 

```suggestion
  ARROW_DCHECK_LE(top_ + alloc_size, buffer_size_)
```